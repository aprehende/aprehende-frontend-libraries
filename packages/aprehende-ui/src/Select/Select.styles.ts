import styled from '@emotion/styled';

export const SelectContainer = styled.div`
  padding: 8px;
  outline: none;
  cursor: pointer;
  border: 1px solid #dbdbdb;
  position: relative;
  width: fit-content;
  border-radius: 3px;
`;

export const ListOfOptions = styled.div<{ isOpen?: boolean }>`
  top: 0;
  left: 0;
  right: 0;
  width: 100%;
  cursor: pointer;
  overflow-y: auto;
  margin-top: 38px;
  min-width: 110px;
  max-height: 180px;
  position: absolute;
  border-radius: 3px;
  background: #ffffff;
  box-sizing: border-box;
  box-shadow: 0 0.5em 1em -0.125em rgba(10, 10, 10, 0.1),
    0 0 0 1px rgba(10, 10, 10, 0.02);
  display: ${({ isOpen }) => (isOpen ? 'block' : 'none')};
  & > * {
    padding: 5px 8px;
  }
  & > *:hover {
    background: rgba(0, 0, 0, 0.05);
  }
`;

export const LabelOptionSelected = styled.div`
  display: flex;
  position: relative;
  color: #0f172a;
  width: fit-content;
  min-width: 110px;
  align-items: center;
  & > div {
    display: block;
    flex: 1;
    margin-right: 8px;
  }
`;

export const LabelOptionHide = styled.div`
  color: #0f172a;
  margin-right: 8px;
  flex: 1;
  min-width: 110px;
`;
