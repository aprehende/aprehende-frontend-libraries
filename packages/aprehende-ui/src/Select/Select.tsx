import { useSelect } from 'downshift';
import { BsChevronDown } from 'react-icons/bs';

import {
  LabelOptionHide,
  LabelOptionSelected,
  ListOfOptions,
  SelectContainer,
} from './Select.styles';

interface SelectProps<T> {
  items: T[];
  renderItem: (item: T) => React.ReactNode;
}

function Select<T>({ items, renderItem }: SelectProps<T>) {
  const {
    isOpen,
    selectedItem,
    getToggleButtonProps,
    getLabelProps,
    getMenuProps,
    getItemProps,
  } = useSelect({
    items,
  });

  return (
    <SelectContainer>
      <LabelOptionSelected
        {...getToggleButtonProps()}
        data-testid="apr-select-button"
      >
        {selectedItem && renderItem(selectedItem)}
        {!selectedItem && (
          <LabelOptionHide {...getLabelProps()}>
            Seleccione un pa√≠s
          </LabelOptionHide>
        )}
        <BsChevronDown />
      </LabelOptionSelected>
      <ListOfOptions
        {...getMenuProps()}
        isOpen={isOpen}
        data-testid="apr-select-menu"
      >
        {isOpen &&
          items.map((item, index) => (
            <div key={index} {...getItemProps({ item, index })}>
              {renderItem(item)}
            </div>
          ))}
      </ListOfOptions>
    </SelectContainer>
  );
}

export default Select;
