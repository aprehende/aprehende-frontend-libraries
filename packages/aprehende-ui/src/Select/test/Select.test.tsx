import { fireEvent, render } from '@testing-library/react';
import { describe, it, vi } from 'vitest';

import Select from '../Select';

interface Country {
  value: string;
  label: string;
}

const countries: Country[] = [
  { value: 'MX', label: 'Mexico' },
  { value: 'US', label: 'United States' },
  { value: 'CA', label: 'Canada' },
];
describe('Select', () => {
  it('should render', () => {
    const { container } = render(
      <Select<Country>
        items={countries}
        renderItem={(item) => <div>{item.label}</div>}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should open the menu when clicking the input', () => {
    const { getByTestId } = render(
      <Select<Country>
        items={countries}
        renderItem={(item) => <div>{item.label}</div>}
      />
    );
    const input = getByTestId('apr-select-button');
    const menu = getByTestId('apr-select-menu');
    fireEvent.click(input);
    expect(menu).toHaveStyle('display: block');
  });

  it('should select an item when clicking on it', () => {
    const { getByText, getAllByRole, getByTestId } = render(
      <Select<Country>
        items={countries}
        renderItem={(item) => <div>{item.label}</div>}
      />
    );
    const input = getByText('Seleccione un país');
    const menu = getByTestId('apr-select-menu');
    fireEvent.click(input);
    expect(menu).toHaveStyle('display: block');
    const item = getAllByRole('option')[0];
    fireEvent.click(item);
    expect(menu).toHaveStyle('display: none');
    expect(getByText('Mexico')).toBeInTheDocument();
  });

  it('should call onChange when selecting an item', () => {
    const onChange = vi.fn();
    const { getByText, getAllByRole, getByTestId } = render(
      <Select<Country>
        items={countries}
        onChange={onChange}
        renderItem={(item) => <div>{item.label}</div>}
      />
    );
    const input = getByText('Seleccione un país');
    const menu = getByTestId('apr-select-menu');
    fireEvent.click(input);
    expect(menu).toHaveStyle('display: block');
    const item = getAllByRole('option')[0];
    fireEvent.click(item);
    expect(menu).toHaveStyle('display: none');
    expect(getByText('Mexico')).toBeInTheDocument();
    expect(onChange).toHaveBeenCalledWith(countries[0]);
  });
});
