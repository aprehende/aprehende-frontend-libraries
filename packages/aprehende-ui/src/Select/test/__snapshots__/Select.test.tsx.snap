// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should render 1`] = `
<div>
  <div
    class="css-9p0mgr"
  >
    <div
      aria-activedescendant=""
      aria-controls="downshift-0-menu"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      class="css-u6w49n"
      data-testid="apr-select-button"
      id="downshift-0-toggle-button"
      role="combobox"
      tabindex="0"
    >
      <div
        class="css-kucoek"
        for="downshift-0-toggle-button"
        id="downshift-0-label"
      >
        Seleccione un pa√≠s
      </div>
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      aria-labelledby="downshift-0-label"
      class="css-no9x8w"
      data-testid="apr-select-menu"
      id="downshift-0-menu"
      role="listbox"
    />
  </div>
</div>
`;
